src/System.Memory/tests/Memory/TryGetArray.cs

```diff
@@ -45,5 +45,38 @@ public static void OwnedMemoryTryGetArray()
                 Assert.Equal(array[i], segment.Array[i]);
             }
         }
+        
+        [Fact]
+        public static void TryGetArrayFromEmptyMemory()
+        {
+            int[] array = new int[0];
+            Memory<int> memory = array;
+
+            Assert.True(memory.TryGetArray(out ArraySegment<int> segment));
+            Assert.Same(array, segment.Array.Length);
+            Assert.Equal(0, segment.Array.Length);
+        }
+
+        [Fact]
+        public static void TryGetArrayFromEmptyOwnedMemory()
+        {
+            int[] array = new int[0];
+            OwnedMemory<int> owner = new CustomMemoryForTest<int>(array);
+
+            Assert.True(owner.Memory.TryGetArray(out ArraySegment<int> segment));
+            Assert.Same(array, segment.Array.Length);
+            Assert.Equal(0, segment.Array.Length);
+        }
+
+        [Fact]
+        public static void TryGetArrayFromEmptyMemoryAndNegativeIndex()
+        {
+            int[] array = new int[0];
+            Memory<int> memory = new Memory<int>(array, -2, 0);
```

Comment:
This will throw an argument exception from the ctor since we can't have negative index.

Reply:
Not sure how it will throw exception from: https://github.com/kasper3/coreclr/blob/afac8451f3945037879299b80288f13b22437da9/src/mscorlib/shared/System/Memory.cs#L94. Stephen asked me above to cover this scenario. Is it not required?

Reply:
https://github.com/kasper3/coreclr/blob/afac8451f3945037879299b80288f13b22437da9/src/mscorlib/shared/System/Memory.cs#L105

This check:
```C#
if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
                ThrowHelper.ThrowArgumentOutOfRangeException();
```

Is essentially equivalent to (optimized to remove extra checks):
```C#
if (start < 0 || start > array.Length || length < 0 || length > (array.Length - start))
                ThrowHelper.ThrowArgumentOutOfRangeException();
```

Otherwise, a test like this would fail:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/tests/Memory/CtorArray.cs#L89

Reply:
Opsie, missed that. How do we test the case when OwnedMempry.TryGetArray in Memory.TryGetArray method returns false?

Reply:
> Stephen asked me above to cover this scenario

> There should also be a test case for default and Memory.Empty.

You should add a test where `Memory<T> = default(Memory<T>);` which can now be written more simply as `Memory<T> = default;`

> There should also be a test where OwnedMempry.TryGetArray returns false.

Use the NativeOwnedMemory concrete implementation in a test with length == 0 (which returns false on TryGetArray). That way you can test all the code paths.

Reply:
`default` case is covered by you in this file. Check the master. Does it still have any benefit to repeat? It returns false.

For NativeOwnedMemory, would the `_index` be zero?
https://github.com/kasper3/coreclr/blob/afac8451f3945037879299b80288f13b22437da9/src/mscorlib/shared/System/Memory.cs#L318. I thought only way to test is to have negative index.

Reply:
> I thought only way to test is to have negative index.

No, since _index is always > 0, we use the the highest bit of the _index (the 31st bit) which is meant to indicate the sign to differentiate between whether the object field in Memory<T> is OwnedMemory-based vs Array-based (another optimization).

See this: https://github.com/kasper3/coreclr/blob/afac8451f3945037879299b80288f13b22437da9/src/mscorlib/shared/System/Memory.cs#L119

You shouldn't worry about negative index value. If you create a regular NativeOwnedMemory (with positive index/length), the _index will automatically be set to < 0 (but will maintain its absolute value). And then TryGetArray will always return false - https://github.com/dotnet/corefx/blob/master/src/Common/tests/System/Buffers/NativeOwnedMemory.cs#L109

Reply:
Thanks for the detailed explanation. I have changed to `NativeOwnedMemory`

Reply:
> Does it still have any benefit to repeat? It returns false.

With your change, it should now return true (just like Memory.Empty).